lib/internal/streams/from.js

```diff
@@ -34,24 +34,52 @@ function from(Readable, iterable, opts) {
     objectMode: true,
     ...opts
   });
+
   // Reading boolean to protect against _read
   // being called before last iteration completion.
   let reading = false;
+
+  // needToClose boolean if iterator needs to be explicitely closed
+  let needToClose = false;
+
   readable._read = function() {
     if (!reading) {
       reading = true;
       next();
     }
   };
+
+  readable._destroy = async function(error, cb) {
+    try {
+      if (needToClose) {
+        needToClose = false;
+        if (typeof iterator.return === 'function') {
+          const { value } = await iterator.return();
+          await value;
```

Comment:
Why is this required? As far as I know `return()` should not provide a `value`?

Reply:
`for-await-of` behaves in such way.
```js
(async () => {
	const deferred = value => new Promise(resolve => setTimeout(() => {
		console.log('resoling with ', value)
		resolve(value)
	}, 500))

	function* generator() {
		yield deferred('first')
		yield deferred('second')
		return deferred('third')
	}

	for await (const x of generator()) {
		console.log('got ', x)
	}
	console.log('done.')
})()
```
output:
```
resoling with  first
got  first
resoling with  second
got  second
resoling with  third
done.
```

Reply:
interesting, good job